src/Artifacts.jl

```diff
@@ -0,0 +1,594 @@
+module Artifacts
+
+import Base: get, SHA1
+import ..depots1
+import ..GitTools: tree_hash, set_readonly
+using ..BinaryPlatforms
+import ..TOML
+import ..Types: parse_toml
+import ..PlatformEngines: download_verify_unpack, probe_platform_engines!, package
+
+export create_artifact, artifact_exists, artifact_path, remove_artifact, verify_artifact,
+       artifact_meta, artifact_hash, bind_artifact, unbind_artifact, download_artifact,
+       find_artifacts_toml, ensure_artifact_installed, @artifact_str, archive_artifact
+
+## Philosophy of Artifacts:
+#
+# - At the lowest level, artifacts are content-addressed chunks of data.  Similarly to
+#   how packages are identified by UUIDs but have human-readable mappings through the
+#   registry, artifacts are identified by their SHA1 git tree hashes, but have human-
+#   readable mappings through `Artifacts.toml` files.
+#
+# - Artifacts do not _need_ to have names bound to them; they are first and foremost
+#   identified by their tree hash.  A name is merely a pointer to a hash, and the hash
+#   itself is what is used in all operations except the initial lookup.
+#
+# - A single name can be associated with multiple hashes, where one is chosen at runtime.
+#   The canonical example of this is platform-dependent artifacts.  At `Artifacts.toml`
+#   parse-time one of the list of hashes will be chosen as the best match for the
+#   current environment (Julia version, OS, processor architecture, etc...), and note
+#   that it is acceptable for there to be multiple possible choices (e.g. an AVX2-
+#   optimized build and a generic x86_64 build that will both run on a modern processor),
+#   however one single hash will be stably chosen from the multiple possibilities.
+#
+# - Artifacts can be created on-machine through the `create()` functional API, and can
+#   also be downloaded as tarballs from external webservers by embedding information into
+#   the `Artifacts.toml` file.  This can be done programmatically through the
+#   `bind_artifact()` function.
```

Comment:
It may be worth mentioning here that Artifacts is not a general purpose download mechanimism,
and it is  as a matter of design philosophy assumed that any artifact hosted will have been created expressly for use with Pkg's artifacts mechanism. (And that it is only a happpy coincidence when it can work on other things)

Reply:
Artifacts should work with any tarball (e.g. `.tar.gz` file).  So it won't work with `.tar.xz` files or `.zip` files, but it works fine with anything else.

Reply:
I don't get why it would work with `.tar.gz` but not `.tar.xz` or `.zip`? If we have the extraction programs available, why wouldn't it work with those?

Reply:
I just mean right now it doesn't work with those.  We can of course include the ability for it to use those extraction programs.